### PR TITLE
Add x-fhir-query enhancement

### LINKED ISSUE #1658: Enhance x-fhir-query to handle fhirpath expression
**Is your feature request related to a problem? Please describe.**
Enhance x-fhir-query to handle fhirpath expression in query as defined in spec https://build.fhir.org/ig/HL7/sdc/expressions.html#x-fhir-query-enhancements

### PR DESCRIPTION
**Pull request by omarismail94**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1658 

**Description**
* Enhance x-fhir-query to handle fhirpath expression in query as defined in spec https://build.fhir.org/ig/HL7/sdc/expressions.html#x-fhir-query-enhancements
* Pass in context into QuestionnaireViewMode
* Small refactor to get rid of multiple `MoreTypes` classes we have

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

In this PR, we pass in the context as a JSON string which we then parse.  I tried passing in the context as  an ID only, so that it can be fetched, but we cannot use the xFhirQueryResolver to fetch the resource as the query  `Resource/Id` does not work

**Type**
Choose one: (Bug fix | **Feature** | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by KhumboLihonga** [comment]
@omarismail94 @jingtang10 Just bringing awareness to this as I see this was marked for review 3 weeks ago but there hasn't been an update since.

**Comment by omarismail94** [comment]
thanks @KhumboLihonga for the ping! We were busy working on the 1.0.0 SDC release and now that's over we can re-focus on this. I updated the PR and will talk to @jingtang10 and team on getting a review

File: datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt [hunk 1]
```diff
@@ -757,7 +757,7 @@ private fun Class<*>.getFieldOrNull(name: String): Field? {
  * objects and throws exception otherwise. This extension function takes care of the conversion
  * based on the input and expected [Type].
  */
-private fun Base.asExpectedType(): Type {
+internal fun Base.asExpectedType(): Type {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/mapping/ResourceMapper.kt:760]: move this to the `extensions` package.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 2]
```diff
@@ -59,6 +59,12 @@ object ExpressionEvaluator {
    */
   private val variableRegex = Regex("[%]([A-Za-z0-9\\-]{1,64})")
 
+  /**
+   * Finds all the matching occurrences of FHIRPaths in x-fhir-query. See:
+   * https://build.fhir.org/ig/HL7/sdc/expressions.html#x-fhir-query-enhancements
+   */
+  internal val xFhirQueryEnhancementRegex = Regex("\\{\\{(.*?)\\}\\}")
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:66]: please add move extensive unit tests to just test this string.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 3]
```diff
@@ -381,6 +385,9 @@ class QuestionnaireFragment : Fragment() {
      */
     internal const val EXTRA_QUESTIONNAIRE_RESPONSE_JSON_STRING = "questionnaire-response"
 
+    /** A JSON encoded string extra for questionnaire context. */
+    const val EXTRA_QUESTIONNAIRE_RESOURCE_CONTEXT_JSON_STRING =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:389]: internal

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 4]
```diff
@@ -493,7 +507,24 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
+
+        val xFhirExpressionString =
+          ExpressionEvaluator.xFhirQueryEnhancementRegex
+            .findAll(expression.expression)
+            .map {
+              val (fhirPathWithParentheses, fhirPath) = it.groupValues
+              fhirPathWithParentheses to
+                fhirPathEngine.evaluate(questionnaireResourceContext, fhirPath).singleOrNull()
+            }
+            .map {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:519]: See if you can use `mapValues` for this one i think

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 5]
```diff
@@ -493,7 +507,24 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
         checkNotNull(xFhirQueryResolver) {
           "XFhirQueryResolver cannot be null. Please provide the XFhirQueryResolver via DataCaptureConfig."
         }
-        xFhirQueryResolver!!.resolve(expression.expression)
+
+        val xFhirExpressionString =
+          ExpressionEvaluator.xFhirQueryEnhancementRegex
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:512]: explore packaging this up in the `ExpressionEvaluator.kt` file in `fhirpath` package

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 6]
```diff
@@ -493,7 +507,24 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
+            .findAll(expression.expression)
+            .map {
+              val (fhirPathWithParentheses, fhirPath) = it.groupValues
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:515]: does this work?

```suggestion
            .map { (fhirPathWithParentheses, fhirPath) ->
```

**Review by jingtang10 — CHANGES_REQUESTED** [review]
👍👍👍

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 4 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: I am not sure if that works. ` Sequence<MatchResult>` is what `findAll` returns and `mapValues`  operates on a `Map<out K, V>`.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 7]
```diff
@@ -513,6 +529,35 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     return item.extractAnswerOptions(data)
   }
 
+  /**
+   * Creates an x-fhir-query from an [Expression]. If the questionnaire resource context is set, the
+   * expression is first checked for any FHIR Path expressions to evaluate first. See:
+   * https://build.fhir.org/ig/HL7/sdc/expressions.html#fhirquery
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:535]:
```suggestion
   * https://build.fhir.org/ig/HL7/sdc/expressions.html#x-fhir-query-enhancements
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 8]
```diff
@@ -513,6 +529,35 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
...
+   */
+  private fun createXFhirQueryFromExpression(expression: Expression): String =
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:537]: optional: maybe organise this functionality in a different package? maybe in a new package called `expression` which includes the current `fhirpath` package.

File: datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt [hunk 9]
```diff
@@ -665,4 +669,95 @@ class ExpressionEvaluatorTest {
...
+
+    val stringBasePairsSequence =
+      ExpressionEvaluator.evaluateXFhirEnhancement(expression, practitioner)
+    assertThat(stringBasePairsSequence.map { it.first }.toList())
+      .containsExactly("{{Practitioner.gender}}")
+
+    assertThat(stringBasePairsSequence.map { it.second }.toList()).containsExactly(null)
+  }
+  @Test
+  fun `evaluateXFhirEnhancement() should  return one pair`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt:740]:
```suggestion
  fun `evaluateXFhirEnhancement() should return one pair`() {
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt [hunk 10]
```diff
@@ -665,4 +669,95 @@ class ExpressionEvaluatorTest {
...
+
+    val stringBasePairsSequence =
+      ExpressionEvaluator.evaluateXFhirEnhancement(expression, practitioner)
+    assertThat(stringBasePairsSequence.map { it.first }.toList()).containsExactly("{{random}}")
+
+    assertThat(stringBasePairsSequence.map { it.second }.toList()).containsExactly(null)
+  }
+
+  @Test
+  fun `evaluateXFhirEnhancement() should evaluate to null for field that does not exist in resource`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt:717]: add a test case where the type of the resource provided to the expression evaluator does not match the base type in the fhir path expression.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
thanks omar - discussed... please see http://build.fhir.org/ig/HL7/sdc/StructureDefinition-sdc-questionnaire-launchContext.html and make sure we're matching the launch context code with the resource provided using the new arg.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt [hunk 11]
```diff
@@ -111,3 +113,34 @@ internal fun StringType.toIdType(): IdType {
...
+ */
+internal fun Base.asExpectedType(): Type {
+  return when (this) {
+    is Enumeration<*> -> toCoding()
+    is IdType -> StringType(idPart)
+    else -> this as Type
+  }
+}
+
+fun Type.valueOrCalculateValue(): Type? {
```
[the remark this answers is not in the document]
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreTypes.kt]: Fixed in this PR

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt (hunk 8 above)
Comment (omarismail94) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: I explored this option and found it just easier to move all the x-fhir stuff in ExpressionEvaluator

**Review by jingtang10 — COMMENTED** [review]
i think the launch context is actually coming from a value set containing a small set of resource types. so please make sure you're enforcing that and that there is documentation.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt [hunk 12]
```diff
@@ -385,6 +389,9 @@ class QuestionnaireFragment : Fragment() {
      */
     internal const val EXTRA_QUESTIONNAIRE_RESPONSE_JSON_STRING = "questionnaire-response"
 
+    /** A JSON encoded string extra for questionnaire context. */
+    internal const val EXTRA_QUESTIONNAIRE_RESOURCE_CONTEXT_JSON_STRING =
+      "questionnaire-resource-id-for-context"
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireFragment.kt:394]:
```suggestion
      "questionnaire-resource-context"
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 13]
```diff
@@ -145,6 +148,22 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
     }
   }
 
+  /** The current resource context as questions are being answered. */
+  private val questionnaireResourceContext: Resource?
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt:152]:
```suggestion
  private val questionnaireLaunchContext: Resource?
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 14]
```diff
@@ -256,6 +252,69 @@ object ExpressionEvaluator {
     return evaluateVariable(expression, questionnaireResponse, variablesMap)
   }
 
+  /**
+   * Create an x-fhir-query string for evaluation
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:256]:
```suggestion
   * Creates an x-fhir-query string for evaluation.
```

and apply for the rest of the PR

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 15]
```diff
@@ -256,6 +252,69 @@ object ExpressionEvaluator {
...
+   *
+   * @param expression x-fhir-query expression
+   * @param resourceContext if passed, the resource context to evaluate the expression against
+   */
+  internal fun createXFhirQueryFromExpression(
+    expression: Expression,
+    resourceContext: Resource?
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:263]:
```suggestion
    launchContext: Resource?
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 16]
```diff
@@ -256,6 +252,69 @@ object ExpressionEvaluator {
...
+        val expressionNode = fhirPathEngine.parse(fhirPath)
+        fhirPathWithParentheses to
+          fhirPathEngine.evaluateToString(
+            mapOf(resource.resourceType.name.lowercase() to resource),
+            null,
+            null,
+            resource,
+            expressionNode
+          )
+      }
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:317]: add new line

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt [hunk 17]
```diff
@@ -256,6 +252,69 @@ object ExpressionEvaluator {
...
+   */
+  internal fun createXFhirQueryFromExpression(
+    expression: Expression,
+    resourceContext: Resource?
+  ): String {
+    if (resourceContext == null) {
+      return expression.expression
+    }
+    return evaluateXFhirEnhancement(expression, resourceContext)
+      .map {
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluator.kt:269]: as discussed, move the warning to `evaluateXFhirEnhancement`

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 18]
```diff
@@ -160,6 +164,192 @@ class QuestionnaireViewModelTest {
       )
   }
 
+  // ==================================================================== //
+  //                                                                      //
+  //        Initialization with Questionnaire Resource Context            //
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:169]:
```suggestion
  //         Initialization with Questionnaire Launch Context             //
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 19]
```diff
@@ -160,6 +164,192 @@ class QuestionnaireViewModelTest {
...
+  //                                                                      //
+  // ==================================================================== //
+  @Test
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:172]: add new line

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 20]
```diff
@@ -160,6 +164,192 @@ class QuestionnaireViewModelTest {
...
+  fun `should throw exception if resource context is set but questionnaire does not contain launchContext extension`() =
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:173]:
```suggestion
  fun `should throw exception if launch context is set but questionnaire does not contain launchContext extension`() =
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 21]
```diff
@@ -160,6 +164,192 @@ class QuestionnaireViewModelTest {
...
+              extension =
+                listOf(
+                  Extension(
+                    "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-answerExpression",
+                    Expression().apply {
+                      this.expression = "Observation?subject={{%patient.id}}"
+                      this.language = Expression.ExpressionLanguage.APPLICATION_XFHIRQUERY.toCode()
+                    }
+                  )
+                )
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:200]: can just remove this

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt [hunk 22]
```diff
@@ -160,6 +164,192 @@ class QuestionnaireViewModelTest {
...
+          .localizedMessage
+
+      assertThat(errorMessage)
+        .isEqualTo(
+          "Resource context set without setting $EXTENSION_SDC_QUESTIONNAIRE_LAUNCH_CONTEXT in the questionnaire."
+        )
+    }
+
+  @Test
+  fun `should throw exception if resource type in context is not part of launchContext set`() =
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:221]: maybe:

```suggestion
  fun `should throw exception if launch context is not one of the allowed resource types`() =
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/QuestionnaireViewModelTest.kt:1]: as discussed, some of these test cases should be in `MoreQuestionnairesTest.kt`

File: datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt [hunk 23]
```diff
@@ -665,4 +672,129 @@ class ExpressionEvaluatorTest {
...
+      }
+
+    val expressionsToEvaluate =
+      ExpressionEvaluator.createXFhirQueryFromExpression(expression, Practitioner())
+
+    assertThat(expressionsToEvaluate).isEqualTo("Practitioner?var1=&var2=&var3=&var4=")
+  }
+
+  @Test
+  fun `createXFhirQueryFromExpression() should evaluate to empty string for incorrect fhir path`() {
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt:693]: this is a degenerative case of the previous

File: datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt [hunk 24]
```diff
@@ -665,4 +672,129 @@ class ExpressionEvaluatorTest {
...
+      Expression().apply {
+        this.language = Expression.ExpressionLanguage.APPLICATION_XFHIRQUERY.toCode()
+        this.expression = "Practitioner?gender={{Practitioner.gender}}"
+      }
+
+    val expressionsToEvaluate =
+      ExpressionEvaluator.createXFhirQueryFromExpression(expression, practitioner)
+    assertThat(expressionsToEvaluate).isEqualTo("Practitioner?gender=")
+  }
+  @Test
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/fhirpath/ExpressionEvaluatorTest.kt:734]: add empty line

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt [hunk 25]
```diff
@@ -107,4 +110,141 @@ class MoreQuestionnairesTest {
...
+            extension =
+              listOf(
+                Extension(
+                  "http://hl7.org/fhir/uv/sdc/StructureDefinition/sdc-questionnaire-answerExpression",
+                  Expression().apply {
+                    this.expression = "Observation?subject={{%patient.id}}"
+                    this.language = Expression.ExpressionLanguage.APPLICATION_XFHIRQUERY.toCode()
+                  }
+                )
+              )
```
Comment (jingtang10) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt:132]: remove this?

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**